universal-application-tool-0.0.1/app/services/applicant/predicate/PredicateEvaluator.java

```diff
@@ -0,0 +1,49 @@
+package services.applicant.predicate;
+
+import services.applicant.ApplicantData;
+import services.applicant.exception.InvalidPredicateException;
+import services.program.predicate.LeafOperationExpressionNode;
+import services.program.predicate.PredicateExpressionNode;
+
+/** Evaluates complex predicates based on the given {@link ApplicantData}. */
+public class PredicateEvaluator {
+
+  private final ApplicantData applicantData;
+  private final JsonPathPredicateGenerator predicateGenerator;
+
+  public PredicateEvaluator(
```

Comment:
(non-blocking) the `JsonPathPredicateGenerator` has `ApplicantQuestion`s, which have `ApplicantData`, and this class also has `ApplicantData`. They may not be the same. Is that weird or nah?

Reply:
I think this should be fine - we're in the context of a single applicant, so the `ApplicantData` should be the same. Additionally, all we needed `ApplicantData` for in the generator is to get the path of the predicate question, which doesn't actually depend on `ApplicantData`.

Reply:
(blocking): if that's the case, can we use `new ApplicantData()` as a useless placeholder in the `JsonPathPredicateGenerator`?

Reply:
Good idea! Added

Reply:
Chatted offline - we're using the new `QuestionDefinition#getContextualizedPath` method here